AssertionError: nodeId out of range: -1 invokeWithArguments(MethodHandle.java:627)

http://jenkins.hazelcast.com/view/freeze/job/freeze-all/13/console

Hazelcast Enterprise 4.1-SNAPSHOT (20200205 - 37bffaf, 1e23c41)

```
02:21:26.259 /disk1/jenkins/workspace/freeze-all/4.0/2020_02_04-11_09_34/stable/destroy/create-use-destroy Failed
02:21:26.265 fail HzClient4HZ _create-use-destroy_createUseDestroy_reliableId hzcmd.id.flake.CreateUseDestroy threadId=0 java.lang.AssertionError: nodeId out of range: -1 [0m
02:21:26.500 http://54.147.27.51/~jenkins/workspace/freeze-all/4.0/2020_02_04-11_09_34/stable/destroy/create-use-destroy
```

```
java.lang.AssertionError: nodeId out of range: -1
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.tryWrapInSameClass(AbstractInvocationFuture.java:1954)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapError(AbstractInvocationFuture.java:1945)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapOrPeel(AbstractInvocationFuture.java:1931)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$ExceptionalResult.wrapForJoinInternal(AbstractInvocationFuture.java:1395)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.resolveAndThrowForJoinInternal(AbstractInvocationFuture.java:592)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.joinInternal(AbstractInvocationFuture.java:576)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.newIdBatch(ClientFlakeIdGeneratorProxy.java:58)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.access$000(ClientFlakeIdGeneratorProxy.java:33)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy$1.newIdBatch(ClientFlakeIdGeneratorProxy.java:45)
	at com.hazelcast.flakeidgen.impl.AutoBatcher.newId(AutoBatcher.java:58)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.newId(ClientFlakeIdGeneratorProxy.java:52)
	at hzcmd.id.flake.CreateUseDestroy.timeStep(CreateUseDestroy.java:11)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

4.1-SNAPSHOT (20200225 - 8b8c625, 4fa4827) 
```
00:10:20.761 /disk1/jenkins/workspace/freeze-x/4.1-SNAPSHOT/2020_02_27-11_38_20/create-use-destroy Failed
00:10:20.769 fail HzClient1HZ _create-use-destroy_createUseDestroy_reliableId hzcmd.id.flake.CreateUseDestroy threadId=0 java.lang.AssertionError: nodeId out of range: -1 [0m
00:10:20.961 http://54.147.27.51/~jenkins/workspace/freeze-x/4.1-SNAPSHOT/2020_02_27-11_38_20/create-use-destroy
```

```
java.lang.AssertionError: nodeId out of range: -1
	at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.tryWrapInSameClass(AbstractInvocationFuture.java:1954)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapError(AbstractInvocationFuture.java:1945)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapOrPeel(AbstractInvocationFuture.java:1931)
	at com.hazelcast.spi.impl.AbstractInvocationFuture$ExceptionalResult.wrapForJoinInternal(AbstractInvocationFuture.java:1395)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.resolveAndThrowForJoinInternal(AbstractInvocationFuture.java:592)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.joinInternal(AbstractInvocationFuture.java:576)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.newIdBatch(ClientFlakeIdGeneratorProxy.java:58)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.access$000(ClientFlakeIdGeneratorProxy.java:33)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy$1.newIdBatch(ClientFlakeIdGeneratorProxy.java:45)
	at com.hazelcast.flakeidgen.impl.AutoBatcher.newId(AutoBatcher.java:58)
	at com.hazelcast.client.impl.proxy.ClientFlakeIdGeneratorProxy.newId(ClientFlakeIdGeneratorProxy.java:52)
	at hzcmd.id.flake.CreateUseDestroy.timeStep(CreateUseDestroy.java:11)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:41)
	at remote.bench.BenchThread.call(BenchThread.java:14)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/hazelcast/src/main/java/com/hazelcast/flakeidgen/impl/FlakeIdGeneratorProxy.java
+++ b/hazelcast/src/main/java/com/hazelcast/flakeidgen/impl/FlakeIdGeneratorProxy.java
@@ -199,7 +199,9 @@ public class FlakeIdGeneratorProxy
      * </ul>
      */
     private int getNodeId() {
-        return getNodeId(System.nanoTime());
+        int nodeId = getNodeId(System.nanoTime());
+        assert nodeId > 0 || nodeId == NODE_ID_OUT_OF_RANGE : "getNodeId() returned invalid value: " + nodeId;
+        return nodeId;
     }
 
     // package-visible for tests
@@ -207,32 +209,35 @@ public class FlakeIdGeneratorProxy
         // Check if it is a time to check for updated nodeId. We need to recheck, because if duplicate node ID
         // is assigned during a network split, this will be resolved after a cluster merge.
         // We throttle the calls to avoid contention due to the lock+unlock call in getMemberListJoinVersion().
-        int nodeId = this.nodeId;
-        if (nodeId != NODE_ID_OUT_OF_RANGE && nextNodeIdUpdate <= nanoTime) {
+        long localNextNodeIdUpdate = this.nextNodeIdUpdate;
+        int localNodeId = this.nodeId;
+        if (localNextNodeIdUpdate <= nanoTime) {
+            if (localNodeId == NODE_ID_OUT_OF_RANGE) {
+                return localNodeId;
+            }
             int newNodeId = getNodeEngine().getClusterService().getMemberListJoinVersion();
             assert newNodeId >= 0 : "newNodeId=" + newNodeId;
             newNodeId += nodeIdOffset;
-            nextNodeIdUpdate = nanoTime + NODE_ID_UPDATE_INTERVAL_NS;
-            if (newNodeId != nodeId) {
-                nodeId = newNodeId;
+            if (newNodeId != localNodeId) {
+                localNodeId = newNodeId;
 
                 // If our node ID is out of range, assign NODE_ID_OUT_OF_RANGE to nodeId
-                if ((nodeId & -1 << bitsNodeId) != 0) {
+                if ((localNodeId & -1 << bitsNodeId) != 0) {
                     outOfRangeMembers.add(getNodeEngine().getClusterService().getLocalMember().getUuid());
-                    logger.severe("Node ID is out of range (" + nodeId + "), this member won't be able to generate IDs. "
-                            + "Cluster restart is recommended.");
-                    nodeId = NODE_ID_OUT_OF_RANGE;
+                    logger.severe("Node ID is out of range (" + localNodeId
+                            + "), this member won't be able to generate IDs. Cluster restart is recommended.");
+                    localNodeId = NODE_ID_OUT_OF_RANGE;
                 }
 
                 // we ignore possible double initialization
-                this.nodeId = nodeId;
+                this.nodeId = localNodeId;
+                this.nextNodeIdUpdate = nanoTime + NODE_ID_UPDATE_INTERVAL_NS;
                 if (logger.isFineEnabled()) {
-                    logger.fine("Node ID assigned to '" + name + "': " + nodeId);
+                    logger.fine("Node ID assigned to '" + name + "': " + localNodeId);
                 }
             }
         }
-
-        return nodeId;
+       return localNodeId;
     }
 
     private Member getRandomMember() {

--- a/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGeneratorProxyTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/flakeidgen/impl/FlakeIdGeneratorProxyTest.java
@@ -41,12 +41,16 @@ import java.net.UnknownHostException;
 import java.util.Date;
 import java.util.Iterator;
 import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
 
 import static com.hazelcast.config.FlakeIdGeneratorConfig.DEFAULT_ALLOWED_FUTURE_MILLIS;
 import static com.hazelcast.config.FlakeIdGeneratorConfig.DEFAULT_BITS_NODE_ID;
 import static com.hazelcast.config.FlakeIdGeneratorConfig.DEFAULT_BITS_SEQUENCE;
 import static com.hazelcast.config.FlakeIdGeneratorConfig.DEFAULT_EPOCH_START;
 import static com.hazelcast.flakeidgen.impl.FlakeIdGeneratorProxy.NODE_ID_UPDATE_INTERVAL_NS;
+import static java.util.concurrent.TimeUnit.SECONDS;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.Mockito.mock;
@@ -72,10 +76,10 @@ public class FlakeIdGeneratorProxyTest {
 
     @Before
     public void before() {
-        before(new FlakeIdGeneratorConfig());
+        initialize(new FlakeIdGeneratorConfig());
     }
 
-    public void before(FlakeIdGeneratorConfig config) {
+    public void initialize(FlakeIdGeneratorConfig config) {
         ILogger logger = mock(ILogger.class);
         clusterService = mock(ClusterService.class);
         NodeEngine nodeEngine = mock(NodeEngine.class);
@@ -105,6 +109,32 @@ public class FlakeIdGeneratorProxyTest {
         assertEquals(20, gen.getNodeId(0));
         assertEquals(20, gen.getNodeId(NODE_ID_UPDATE_INTERVAL_NS - 1));
         assertEquals(30, gen.getNodeId(NODE_ID_UPDATE_INTERVAL_NS));
+    }
+
+    @Test
+    public void givenNodeIdUninitialized_whenNodeIdRequestedConcurrently_thenItNeverReturnUninitializedId() throws Exception {
+        when(clusterService.getMemberListJoinVersion()).thenReturn(20);
+        int threadCount = 20;
+        int iterationCount = 5_000;
+
+        ExecutorService executorService = Executors.newFixedThreadPool(threadCount);
+        AtomicInteger errorCounter = new AtomicInteger();
+        FlakeIdGeneratorConfig genConfig = new FlakeIdGeneratorConfig();
+        for (int i = 0; i < iterationCount; i++) {
+            initialize(genConfig);
+            FlakeIdGeneratorProxy localGen = gen;
+            Runnable getNodeId = () -> {
+                if (localGen.getNodeId(0) == -1) { // see FlakeIdGeneratorProxy#NODE_ID_NOT_YET_SET
+                    errorCounter.incrementAndGet();
+                }
+            };
+            for (int z = 0; z < threadCount; z++) {
+                executorService.submit(getNodeId);
+            }
+        }
+        executorService.shutdown();
+        executorService.awaitTermination(30, SECONDS);
+        assertEquals(0, errorCounter.get());
     }
 
     @Test
@@ -180,7 +210,7 @@ public class FlakeIdGeneratorProxyTest {
     public void test_positiveNodeIdOffset() {
         int nodeIdOffset = 5;
         int memberListJoinVersion = 20;
-        before(new FlakeIdGeneratorConfig().setNodeIdOffset(nodeIdOffset));
+        initialize(new FlakeIdGeneratorConfig().setNodeIdOffset(nodeIdOffset));
 
         when(clusterService.getMemberListJoinVersion()).thenReturn(memberListJoinVersion);
         assertEquals((memberListJoinVersion + nodeIdOffset), gen.getNodeId(0));
@@ -190,7 +220,7 @@ public class FlakeIdGeneratorProxyTest {
     public void when_customBits_then_used() {
         int bitsSequence = 10;
         int bitsNodeId = 11;
-        before(new FlakeIdGeneratorConfig()
+        initialize(new FlakeIdGeneratorConfig()
                 .setBitsSequence(bitsSequence)
                 .setBitsNodeId(bitsNodeId)
                 .setEpochStart(0));
@@ -205,7 +235,7 @@ public class FlakeIdGeneratorProxyTest {
     public void when_epochStart_then_used() {
         int epochStart = 456;
         int timeSinceEpochStart = 1;
-        before(new FlakeIdGeneratorConfig().setEpochStart(epochStart));
+        initialize(new FlakeIdGeneratorConfig().setEpochStart(epochStart));
         long id = gen.newIdBaseLocal(epochStart + timeSinceEpochStart, 1234, 10).idBatch.base();
         assertEquals((timeSinceEpochStart << DEFAULT_BITS_SEQUENCE + DEFAULT_BITS_NODE_ID) + 1234, id);
     }

